packages/node/src/exec/api.ts

```diff
@@ -0,0 +1,106 @@
+import { Operation } from 'effection';
+import { Channel } from '@effection/channel';
+
+/**
+ * The process type is what is returned by the `exec` operation. It has all of
+ * standard io handles, and methods for synchronizing on return.
+ */
+export interface Process extends StdIO {
+  /**
+   * Completes once the process has finished regardless of whether it was
+   * successful or not.
+   */
+  join(): Operation<ExitStatus>;
+
+  /**
+   * Completes once the process has finished successfully. If the process does
+   * not complete successfully, it will raise an ExecError.
+   */
+  expect(): Operation<ExitStatus>;
+}
+
+export interface ExecOptions {
+  /**
+   * When not using passing the `shell` option all arguments must be passed
+   * as an array.
+   */
+  arguments?: string[];
+
+  /**
+   * Map of environment variables to use for the process.
+   */
+  env?: Record<string, string>;
+
+  /**
+   * Create an intermediate shell process. Useful if you need to handle glob
+   * expantion;
+   */
+  shell?: boolean;
+
+  /**
+   * Sets the working directory of the process
+   */
+  cwd?: string;
+}
+
+export interface StdIO {
+  stdout: Channel<string>;
+  stderr: Channel<string>;
+  stdin: Channel<string>;
+}
+
+export interface ExitStatus {
+  /**
+   * exit code
+   * //TODO: is this pertinent on Windows? Do we need an 'OK' flag
+   */
+  code?: number;
+
+  /**
+   * If the process exited with a signal instead of an exit code, it
+   * is recorded here.
+   */
+  signal?: string;
+
+  /**
+   * A buffer of the recent output from both `stdout` and `stderr` that can
+   * be helpful for debugging
+   */
+  tail: string[];
+
+  /**
+   * The original command used to create the process
+   */
+  command: string;
+
+  /**
+   * The original options used to create the process.
+   */
+  options: ExecOptions;
```

Comment:
How about moving these to the process, and instead providing a reference to the process in the exit status. So we can do `exitStatus.process.options`, this should allow even more flexibility.

Reply:
The main issue I'd have with that is that it would convert what amounts to a simple value object into a holder for the stateful process object. I'm doubly nervous about hanging a live resource off of an object that could be handed around to other functions that could easily run outside the scope of the resource's context.

Reply:
Fair point!